src/Domain/WeightedElementDistribution.hpp

```diff
@@ -0,0 +1,147 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+template <size_t Dim>
+class Block;
+
+template <size_t Dim>
+class ElementId;
+
+namespace Spectral {
+enum class Quadrature;
+}  // namespace Spectral
+
+namespace domain {
+
+/*!
+ * \brief Distribution strategy for assigning elements to CPUs using a
+ * Morton ('Z-order') space-filling curve to determine placement within each
+ * block, where elements are assigned weighted costs that are distributed
+ * across CPUs
+ *
+ * \details The element distribution attempts to assign a balanced total
+ * computational cost to each processor that is allowed to have elements. The
+ * cost assigned to a processor is defined as the sum of the costs of the
+ * `Element`s assigned to a processor, and the cost of an `Element` is defined
+ * as `(number of grid points) / sqrt(minimum grid spacing in Frame::Grid)`.
+ * First, each `Block`'s `Element`s are ordered by their Z-curve index (see more
+ * below). `Element`s are traversed in this order and assigned to CPUs in order,
+ * moving onto the next CPU once the target cost per CPU is met. The target cost
+ * per CPU is defined as the remaining cost to distribute divided by the
+ * remaining number of CPUs to distribute to. This is an important distinction
+ * from simply having one constant target cost per CPU defined as the total cost
+ * divided by the total number of CPUs with elements. Since elemental costs will
```

Comment:
what is an elemental cost?

Reply:
`elemental cost` just refers to the computed cost of an `Element`. Since I previously defined how the cost is computed, I'm guessing your question means that `elemental costs` as a phrase wasn't clearly referring to this costs of an `Element`. Would it be more clear if I said `Since the cost of Elements on a processor` or am I misinterpreting your question?

Reply:
Yea, you've got it! Yes, `Since the total cost of Elements on a...` would've cleared up my confusion!